File: elixir/sum-of-multiples/sum_of_multiples.exs
defmodule SumOfMultiples do
  @doc """
  Adds up all numbers from 1 to a given end number that are multiples of the factors provided.

  The default factors are 3 and 5.
  """
  @spec to(non_neg_integer, [non_neg_integer]) :: non_neg_integer
  def to(limit, factors \\ [3, 5]) do
    1..limit - 1
    |> Enum.filter(fn x -> factor?(x, factors) end)
    |> Enum.sum
  end

  def factor?(n, factors) do
    factors |> Enum.any?(fn x -> rem(n, x) == 0 end)
  end
end

File: erlang/sum-of-multiples/sum_of_multiples.erl
-module(sum_of_multiples).
-export([sumOfMultiplesDefault/1, sumOfMultiples/2]).

sumOfMultiples(Factors, N) ->
    lists:sum([X || X <- lists:seq(1, N-1), hasFactor(X, Factors)]).

sumOfMultiplesDefault(N) ->
    sumOfMultiples([3, 5], N).

hasFactor(X, Factors) ->
    lists:any(fun (Factor) -> X rem Factor =:= 0 end, Factors).
